solidity/ecdsa/contracts/libraries/DKG.sol

```diff
@@ -0,0 +1,509 @@
+//   ▓▓▓▓▓▓▄▄▓▓▓▓▓▓▓▀      ▐▓▓▓▓▓▓▄▄▄▄         ▓▓▓▓▓▓▄▄▄▄         ▐▓▓▓▓▓▌   ▐▓▓▓▓▓▓
+//   ▓▓▓▓▓▓▓▓▓▓▓▓▓▀        ▐▓▓▓▓▓▓▓▓▓▓         ▓▓▓▓▓▓▓▓▓▓         ▐▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓
+//   ▓▓▓▓▓▓▀▀▓▓▓▓▓▓▄       ▐▓▓▓▓▓▓▀▀▀▀         ▓▓▓▓▓▓▀▀▀▀         ▐▓▓▓▓▓▓▓▓▓▓▓▓▓▓▀
+//   ▓▓▓▓▓▓   ▀▓▓▓▓▓▓▄     ▐▓▓▓▓▓▓     ▓▓▓▓▓   ▓▓▓▓▓▓     ▓▓▓▓▓   ▐▓▓▓▓▓▌
+// ▓▓▓▓▓▓▓▓▓▓ █▓▓▓▓▓▓▓▓▓ ▐▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓ ▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓  ▓▓▓▓▓▓▓▓▓▓
+// ▓▓▓▓▓▓▓▓▓▓ ▓▓▓▓▓▓▓▓▓▓ ▐▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓ ▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓  ▓▓▓▓▓▓▓▓▓▓
+//
+//
+
+// Initial version copied from Keep Network Random Beacon:
+// https://github.com/keep-network/keep-core/blob/5138c7628868dbeed3ae2164f76fccc6c1fbb9e8/solidity/random-beacon/contracts/libraries/DKG.sol
+//
+// With a difference in group size which was set to 100.
+
+pragma solidity ^0.8.9;
+
+import "@openzeppelin/contracts/utils/cryptography/ECDSA.sol";
+import "@keep-network/sortition-pools/contracts/SortitionPool.sol";
+import "./BytesLib.sol";
+import "./Submission.sol";
+import "../DKGValidator.sol";
+
+library DKG {
+    using BytesLib for bytes;
+    using ECDSA for bytes32;
+
+    struct Parameters {
+        // Time in blocks during which a submitted result can be challenged.
+        uint256 resultChallengePeriodLength;
+        // Time in blocks after which the next group member is eligible
+        // to submit DKG result.
+        uint256 resultSubmissionEligibilityDelay;
+    }
+
+    struct Data {
+        // Address of the Sortition Pool contract.
+        SortitionPool sortitionPool;
+        // Address of the DKGValidator contract.
+        DKGValidator dkgValidator;
+        // DKG parameters. The parameters should persist between DKG executions.
+        // They should be updated with dedicated set functions only when DKG is not
+        // in progress.
+        Parameters parameters;
+        // Time in blocks at which DKG started.
+        uint256 startBlock;
+        // Seed used to start DKG.
+        uint256 seed;
+        // Time in blocks that should be added to result submission eligibility
+        // delay calculation. It is used in case of a challenge to adjust
+        // block calculation for members submission eligibility.
+        uint256 resultSubmissionStartBlockOffset;
+        // Hash of submitted DKG result.
+        bytes32 submittedResultHash;
+        // Block number from the moment of the DKG result submission.
+        uint256 submittedResultBlock;
+    }
+
+    /// @notice DKG result.
+    struct Result {
+        // Claimed submitter candidate group member index.
+        // Must be in range [1, groupSize].
+        uint256 submitterMemberIndex;
+        // Generated candidate group public key
+        bytes groupPubKey;
+        // Array of misbehaved members indices (disqualified or inactive).
+        // Indices must be in range [1, groupSize], unique, and sorted in ascending
+        // order.
+        uint8[] misbehavedMembersIndices;
```

Comment:
Since this is still pre-audit change, wdyt about modifying it to `uint256[]`? In random-beacon tests, it shows that by changing to `uint256[]` we can save around ~15k when calling `submitDkgResult` and ~10k for `addCandidateGroup`. This can be a separate PR.

Reply:
Sure, I'll compare the costs and push it in this PR.

Reply:
Unfortunatelly I didn't notice a significant gas cost reduction after updating to  `uint256[] misbehavedMembersIndices`.


### uint8[] misbehavedMembersIndices

```
·--------------------------------------------|---------------------------|-------------|-----------------------------·
|            Solc version: 0.8.9             ·  Optimizer enabled: true  ·  Runs: 200  ·  Block limit: 30000000 gas  │
·············································|···························|·············|······························
|  Methods                                                                                                           │
···················|·························|·············|·············|·············|···············|··············
|  Contract        ·  Method                 ·  Min        ·  Max        ·  Avg        ·  # calls      ·  eur (avg)  │
···················|·························|·············|·············|·············|···············|··············
|  SortitionPool   ·  insertOperator         ·     192273  ·     329012  ·     196963  ·           67  ·          -  │
···················|·························|·············|·············|·············|···············|··············
|  SortitionPool   ·  lock                   ·          -  ·          -  ·      26494  ·            1  ·          -  │
···················|·························|·············|·············|·············|···············|··············
|  StakingStub     ·  increaseAuthorization  ·      44858  ·      44870  ·      44869  ·          132  ·          -  │
···················|·························|·············|·············|·············|···············|··············
|  WalletRegistry  ·  approveDkgResult       ·     235231  ·     265610  ·     249961  ·           38  ·          -  │
···················|·························|·············|·············|·············|···············|··············
|  WalletRegistry  ·  challengeDkgResult     ·     212702  ·    1685115  ·    1216517  ·           24  ·          -  │
···················|·························|·············|·············|·············|···············|··············
|  WalletRegistry  ·  notifyDkgTimeout       ·      43603  ·      44803  ·      43903  ·            4  ·          -  │
···················|·························|·············|·············|·············|···············|··············
|  WalletRegistry  ·  registerOperator       ·     201670  ·     338409  ·     206430  ·          132  ·          -  │
···················|·························|·············|·············|·············|···············|··············
|  WalletRegistry  ·  requestNewWallet       ·          -  ·          -  ·      82486  ·           22  ·          -  │
···················|·························|·············|·············|·············|···············|··············
|  WalletRegistry  ·  requestSignature       ·          -  ·          -  ·      50161  ·            4  ·          -  │
···················|·························|·············|·············|·············|···············|··············
|  WalletRegistry  ·  submitDkgResult        ·     270853  ·     307443  ·     299882  ·           65  ·          -  │
···················|·························|·············|·············|·············|···············|··············
|  WalletRegistry  ·  submitSignature        ·          -  ·          -  ·      37173  ·            3  ·          -  │
···················|·························|·············|·············|·············|···············|··············
|  WalletRegistry  ·  updateDkgParams        ·          -  ·          -  ·      40601  ·            2  ·          -  │
···················|·························|·············|·············|·············|···············|··············
|  Deployments                               ·                                         ·  % of limit   ·             │
·············································|·············|·············|·············|···············|··············
|  DKGValidator                              ·          -  ·          -  ·    1432761  ·        4.8 %  ·          -  │
·············································|·············|·············|·············|···············|··············
|  SortitionPool                             ·          -  ·          -  ·    2245461  ·        7.5 %  ·          -  │
·············································|·············|·············|·············|···············|··············
|  StakingStub                               ·          -  ·          -  ·     346156  ·        1.2 %  ·          -  │
·--------------------------------------------|-------------|-------------|-------------|---------------|-------------·
```

### uint256[] misbehavedMembersIndices

```
·--------------------------------------------|---------------------------|-------------|-----------------------------·
|            Solc version: 0.8.9             ·  Optimizer enabled: true  ·  Runs: 200  ·  Block limit: 30000000 gas  │
·············································|···························|·············|······························
|  Methods                                                                                                           │
···················|·························|·············|·············|·············|···············|··············
|  Contract        ·  Method                 ·  Min        ·  Max        ·  Avg        ·  # calls      ·  eur (avg)  │
···················|·························|·············|·············|·············|···············|··············
|  SortitionPool   ·  insertOperator         ·     192273  ·     329012  ·     196963  ·           67  ·          -  │
···················|·························|·············|·············|·············|···············|··············
|  SortitionPool   ·  lock                   ·          -  ·          -  ·      26494  ·            1  ·          -  │
···················|·························|·············|·············|·············|···············|··············
|  StakingStub     ·  increaseAuthorization  ·      44858  ·      44870  ·      44869  ·          132  ·          -  │
···················|·························|·············|·············|·············|···············|··············
|  WalletRegistry  ·  approveDkgResult       ·     237115  ·     263618  ·     251471  ·           38  ·          -  │
···················|·························|·············|·············|·············|···············|··············
|  WalletRegistry  ·  challengeDkgResult     ·     210505  ·    1684896  ·    1217005  ·           24  ·          -  │
···················|·························|·············|·············|·············|···············|··············
|  WalletRegistry  ·  notifyDkgTimeout       ·      43603  ·      44803  ·      43903  ·            4  ·          -  │
···················|·························|·············|·············|·············|···············|··············
|  WalletRegistry  ·  registerOperator       ·     201670  ·     338409  ·     206430  ·          132  ·          -  │
···················|·························|·············|·············|·············|···············|··············
|  WalletRegistry  ·  requestNewWallet       ·          -  ·          -  ·      82508  ·           22  ·          -  │
···················|·························|·············|·············|·············|···············|··············
|  WalletRegistry  ·  requestSignature       ·          -  ·          -  ·      50161  ·            4  ·          -  │
···················|·························|·············|·············|·············|···············|··············
|  WalletRegistry  ·  submitDkgResult        ·     272157  ·     306623  ·     301459  ·           65  ·          -  │
···················|·························|·············|·············|·············|···············|··············
|  WalletRegistry  ·  submitSignature        ·          -  ·          -  ·      37161  ·            3  ·          -  │
···················|·························|·············|·············|·············|···············|··············
|  WalletRegistry  ·  updateDkgParams        ·          -  ·          -  ·      40512  ·            2  ·          -  │
···················|·························|·············|·············|·············|···············|··············
|  Deployments                               ·                                         ·  % of limit   ·             │
·············································|·············|·············|·············|···············|··············
|  DKGValidator                              ·          -  ·          -  ·    1392593  ·        4.6 %  ·          -  │
·············································|·············|·············|·············|···············|··············
|  SortitionPool                             ·          -  ·          -  ·    2245461  ·        7.5 %  ·          -  │
·············································|·············|·············|·············|···············|··············
```

Reply:
ah, sorry, I think I was on the non updated branch, cause now when I double checked I see that the only significant difference in random-beacon is in `challengeDkgResult`, which makes sense. We moved members hash to `challengeDkgResult` which uses `misbehavedMembersIndices`

uint256
```
challengeDkgResult  337043  ·    1272758  ·    1006948
```

uint8[]
```
challengeDkgResult  336904  ·    1288261  ·    1009187
```

Diff `15503`. Other functions showed fluctuation between 0~2000

Not sure why there's not that much difference in /ecdsa for `challengeDkgResult`. Anyway, I think we can leave the type as-is or change it to uin256[], your call.